File: Problem3.java
// Time Complexity - mlogn 
// Space Complexity - O(1)
// Executed the code in local SDE

import java.util.*;
public class Optimize_Air_Routes {
	public static List<Integer> result = new ArrayList<>();
	public static void main(String args[]) {
		int target = 10000;
		int[][] forwardRouteList = {{1,3000},{2,5000},{3,7000},{4,10000}};
		int[][] returnRouteList = {{1,2000},{2,3000},{3,4000},{4,5000}};
		SearchPairs(forwardRouteList,returnRouteList,target);
		System.out.println(result.toString());
		
	}
	private static void SearchPairs(int[][] forwardRouteList,int[][] returnRouteList,int target) {
		int max =0;
		for(int i=0;i < forwardRouteList.length;i++ ) {
			int compliment = target - forwardRouteList[i][1]; // find the compliment of the target which would be the target for the binary search in returnRouteList
			int BackIndx = BinarySearch(returnRouteList, compliment); // The BinarySearch will return the index where the closest value to compliment was found
			if(BackIndx >= 0) { // Edge Case when valid BackIndex is not found
				int sum = forwardRouteList[i][1] + returnRouteList[BackIndx][1]; // Calculate the sum of forwardRoute distance and closest Backward route distance
				if(sum > max) {	// if the sum is greater than the max. add the route index to the result						
					result = new ArrayList<>();
					max = sum;
					result.add(forwardRouteList[i][0]);
					result.add(returnRouteList[BackIndx][0]);
				}	
			}
		}
	}
	
	private static int BinarySearch(int[][] arr, int target ) {		//BinarySearch to find the closest distance in returnroute for the complimment
		int low = 0;
		int high = arr.length -1;
		while(low <= high) {
			int mid = low + (high - low)/2; // To avoid Integer overflow
			if(arr[mid][1] == target)	return mid;
			else if(arr[mid][1] < target)	low = mid +1;
			else high = mid-1;
		}
		return low -1; // if mid is not equal to the target, return the closest distance less than the target
	}
}

File: Problem3.py
class Solution:
	def OptimizedAir1(self, maxTravelDist, forwardRouteList, returnRouteList):
		# Brute Force Way
		# Using more space for storing all the differences
		mapDistance = dict()
		for i in forwardRouteList:
			if i[1] > maxTravelDist:
				continue
			for j in returnRouteList:
				if j[1] > maxTravelDist - i[1]:
					continue
				totalDist = maxTravelDist - (i[1] + j[1])
				if totalDist not in mapDistance:
					mapDistance[totalDist] = []
				mapDistance[totalDist].append((i[0], j[0]))
		sortedKeys = sorted(list(mapDistance.keys()))
		for i in sortedKeys:
			if i < 0:
				continue
			return mapDistance[i]

	def OptimizedAir2(self, maxTravelDist, forwardRouteList, returnRouteList):
		# Brute force way
		# Time Complexity : O(nm) where n is the length of the forwardRouteList and m is the length of returnRouteList
		maxDist = -1
		maxDistlist = []
		for i in forwardRouteList:
			if i[1] > maxTravelDist:
				continue
			for j in returnRouteList:
				if j[1] > maxTravelDist - i[1]:
					continue
				totalDist = i[1] + j [1]
				if totalDist >= maxDist:
					if totalDist > maxDist:
						maxDistlist = []
						maxDist = totalDist
					maxDistlist.append((i[0], j[0]))
		return maxDistlist

	def OptimizedAir3(self, maxTravelDist, forwardRouteList, returnRouteList):
		# Using the binary search to find the nearest value to the compliment
		# Time Complexity : O(nlogn)[sort forwardRouteList] + O(mlogm)[sort returnRouteList] + O(nlogm) [Iterating through the forwardRouteList and finding the closest value to the compliment from the returnRouteList using binary search]
		forwardRouteList.sort(key = lambda x:x[1])
		returnRouteList.sort(key = lambda x:x[1])
		maxDist = -1
		maxDistlist = []
		for i in forwardRouteList:
			compliment = maxTravelDist - i[1]
			nearest = self.getNearestDist(returnRouteList, compliment)
			if nearest != -1:
				totalDist = i[1] + returnRouteList[nearest][1]
				if totalDist >= maxDist:
					if totalDist > maxDist:
						maxDistlist = []
						maxDist = totalDist
					maxDistlist.append((i[0], returnRouteList[nearest][0]))
		return maxDistlist

	def getNearestDist(self, arr, target):
		low = 0
		high = len(arr) - 1
		nearest = -1
		while low <= high:
			mid = low + (high - low)//2
			if arr[mid][1] == target:
				return mid
			elif arr[mid][1] > target:
				high = mid - 1
			else:
				nearest = mid
				low = mid + 1
		return nearest

if __name__ == '__main__':
	oa = Solution()
	print(oa.OptimizedAir1(7000, [[1,2000],[2,4000],[3,6000]], [[1,2000]]))
	print(oa.OptimizedAir1(10000, [[1,3000],[2,5000],[3,7000],[4,10000]], [[1,2000],[2,3000],[3,4000],[4,5000]]))
	print(oa.OptimizedAir2(7000, [[1,2000],[2,4000],[3,6000]], [[1,2000]]))
	print(oa.OptimizedAir2(10000, [[1,3000],[2,5000],[3,7000],[4,10000]], [[1,2000],[2,3000],[3,4000],[4,5000]]))
	print(oa.OptimizedAir3(7000, [[1,2000],[2,4000],[3,6000]], [[1,2000]]))
	print(oa.OptimizedAir3(10000, [[1,3000],[2,5000],[3,7000],[4,10000]], [[1,2000],[2,3000],[3,4000],[4,5000]]))